cql3/statements/select_statement.cc

```diff
@@ -543,14 +544,16 @@ indexed_table_select_statement::do_execute_base_query(
         auto &merger = query_state.merger;
         auto &keys = query_state.primary_keys;
         auto &key_it = query_state.current_primary_key;
-        return repeat([this, &keys, &key_it, &merger, &proxy, &state, &options, cmd, timeout]() {
+        return repeat([this, &keys, &key_it, &merger, &proxy, &state, &options, cmd, timeout, &query_state]() {
             // Starting with 1 key, we check if the result was a short read, and if not,
             // we continue exponentially, asking for 2x more key than before
             auto key_it_end = std::min(key_it + std::distance(keys.begin(), key_it) + 1, keys.end());
             auto command = ::make_lw_shared<query::read_command>(*cmd);
 
             query::result_merger oneshot_merger(cmd->row_limit, query::max_partitions);
-            return map_reduce(key_it, key_it_end, [this, &proxy, &state, &options, cmd, timeout] (auto& key) {
+            return map_reduce(key_it, key_it_end, [this, &proxy, &state, &options, cmd, timeout, &query_state] (auto& key) {
+              // Limit concurrency here to avoid flooding the reader_concurrency_semaphore (and in general generating huge loads). See #3807.
+              return with_semaphore(query_state.concurrency_limiter, 1, [this, &proxy, &state, &options, cmd, timeout, key] {
```

Comment:
if I understand correctly what this patch does, it seems good even outside of the context of the reader_concurrency_semaphore.